README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://CRAN.R-project.org/web/checks/check_results_startup.html"><img border="0" src="https://www.r-pkg.org/badges/version/startup" alt="CRAN check status"/></a> <a href="https://github.com/HenrikBengtsson/startup/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/startup/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/startup"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/startup/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# startup: Friendly R Startup Configuration 

## Introduction

When you start R, it will by default source a `.Rprofile` file if it exists.  This allows you to automatically tweak your R settings to meet your everyday needs.  For instance, you may want to set the default CRAN repository (`options("repos")`) so you don't have to choose one every time you install a package.

The **[startup]** package extends the default R startup process by allowing you to put multiple startup scripts in a common 'Rprofile.d' directory and have them all be sourced during the R startup process.  This way you can have one file to configure the default CRAN repository and another one to configure your personal devtools settings.
Similarly, you can use a 'Renviron.d' directory with multiple files defining different environment variables.  For instance, one file may define environment variable `LANGUAGE`, whereas another file may contain your private `GITHUB_PAT` key.
The advantages of this approach are that it gives a better overview when you list the files, makes it easier to share certain settings (= certain files) with other users, and enable you to keep specific files completely private (by setting the file privileges so only you can access those settings).


## How the R startup process works

When R starts, the following _user-specific_ setup takes place:

1. The _first_ `.Renviron` file found on the R startup search path is processed.  The search path is (in order): (i) `Sys.getenv("R_ENVIRON_USER")`, (ii) `./.Renviron`, and (iii) `~/.Renviron`.  The format of this file is one `ENV_VAR=VALUE` statement per line, cf. `?.Renviron`.  _NOTE:_ Some environment variables must be set already in this step in order to be acknowledged by R, i.e. it is _too late_ to set some of them in Step 2 and Step 3 below.

2. The _first_ `.Rprofile` file found on the R startup search path is processed.  The search path is (in order): (i) `Sys.getenv("R_PROFILE_USER")`, (ii) `./.Rprofile`, and (iii) `~/.Rprofile`.  The format of this file must be a valid R script (with a trailing newline), cf. `?.Rprofile`.

3. If the `.Rprofile` file (in Step 2) calls `startup::startup()` then the following will also take place:

   a. The _first_ 'Renviron.d' directory on the R startup search path is processed.  The search path is (in order): (i) `paste0(Sys.getenv("R_ENVIRON_USER"), ".d")`, (ii) `./.Renviron.d`, (iii) `~/.Renviron.d`, and (iv) `{user-config-dir}/Renviron.d`, where `{user-config-dir}` corresponds to `tools::R_user_dir("startup", which = "config")`, e.g. `${XDG_CONFIG_HOME}/R/startup`.  The format of these files should be the same as for `.Renviron`.  _NOTE:_ Some environment variables must be set already in Step 1 above in order to be acknowledged by R.

   b. A set of handy R options that can be used in Step 3c are set.  Their names are prefixed `startup.session.` - see `?startup::startup_session_options` for details.

   c. The _first_ 'Rprofile.d' directory found on the R startup search path is processed.  The search path is (in order): (i) `paste0(Sys.getenv("R_PROFILE_USER"), ".d")`, (ii) `./.Rprofile.d`, (iii) `~/.Rprofile.d`, and (iv) `{user-config-dir}/Rprofile.d`.  The format of these files should be the same as for `.Rprofile`, that is, they must be valid R scripts.

   d. If set, any R code in environment variable `R_STARTUP_INIT`, or R option `startup.init`, is parsed and evaluated.
   
   e. If set, any R script specified by environment variable `R_STARTUP_FILE`, or R option `startup.file`, is parsed and evaluated.
   
   f. If no errors occur above, the **[startup]** package will be unloaded, leaving no trace of itself behind, except for R options `startup.session.*` set in Step 3b - these will be erased if `startup::startup()` is called with `keep = NULL`.


All relevant files in 'Renviron.d' and 'Rprofile.d'  directories, including those found recursively in subdirectories thereof, will be processed (in lexicographic order sorted under the `C` locale).
There are no restrictions on what the file names should be (except for the ones ignored as explained below).  However, for 'Rprofile.d' files we recommend to use filename extension `*.R` to indicate that the files are regular R scripts.  For 'Renviron.d' files we recommend to use files without extensions (or `*.Renviron` for clarification).  To avoid confusions, do _not_ use an `*.R` extension for 'Renviron.d' files because they are not R script per se (as some editors may warn you about).

Files with file extensions `*.txt`, `*.md` and `*~` are ignored as well as any files named `.Rhistory`, `.RData` and `.DS_Store`.  Directories named `__MACOSX` and their content are ignored.  Files and directories with names starting with two periods (`..`) are ignored, e.g. `~/.Rprofile.d/..my-tests/`.



## Installation

After installing the **startup** packages (see instructions at the end), call

```r
startup::install()
```

once.  This will append

```r
tryCatch(startup::startup(), error=function(ex) message(".Rprofile error: ", conditionMessage(ex)))
```

to your `~/.Rprofile`.  The file will be created if missing.  This will also create directories `~/.Renviron.d/` and `~/.Rprofile.d/` if missing.  To find the location of these folder on Windows, use `normalizePath("~")` - it's often located under `C:\Users\Alice\Documents\`.


Alternatively to the above installation setup, you can just add that line to your `~/.Rprofile` file manually.  The reason for using `tryCatch(..., error = ...)` is for the case when **startup** is not installed and you try to install it, e.g. after upgrading R to a new major release.  Without `tryCatch()`, R will fail to install the **startup** package (or any other package) because the R profile startup script produces an error complaining about **startup** not being available.




## Usage

Just start R :)

To debug the startup process, use `startup::startup(debug = TRUE)` or set environment variable `R_STARTUP_DEBUG=TRUE`, e.g. on Linux you can do

```sh
$ R_STARTUP_DEBUG=TRUE R
```

On MS Windows, you can do:

```sh
$ R R_STARTUP_DEBUG=TRUE
```

This will produce time-stamped messages during startup specifying which files are included together with details on the files and what they modified.


## Conditional file and directory names

If the name of a file consists of a `<key>=<value>` specification, then that file will be included / used only if the specification is fulfilled (on the current system with the current R setup).  For instance, a file `~/.Rprofile.d/os=windows.R` will be ignored unless `startup::sysinfo()$os == "windows"`, i.e. the R session is started on a Windows system.

The following `startup::sysinfo()` keys are available for conditional inclusion of files by their path names:

* System values:
  - `dirname`     - (character) the name of the current working directory (= `basename(getwd())`)
  - `gui`         - (character) the graphical user interface (= `.Platform$GUI`)
  - `nodename`    - (character) the host name (= `Sys.info()[["nodename"]]`)
  - `machine`     - (character) the machine type (= `Sys.info()[["machine"]]`)
  - `os`          - (character) the operating system (= `.Platform$OS.type`)
  - `sysname`     - (character) the system name (= `Sys.info()[["sysname"]]`)
  - `user`        - (character) the user name (= `Sys.info()[["user"]]`)
  
* System flags:
  - `interactive` - (logical) whether running R interactively or not (= `interactive()`)
  - `ess`         - (logical) whether running R in [Emacs Speaks Statistics (ESS)] or not
  - `pqr`         - (logical) whether running [pqR] ("A Pretty Quick Version of R") or not
  - `radian`      - (logical) whether running R in [radian] (formerly known as 'rtichoke' and 'rice') or not
  - `microsoftr`  - (logical) whether running R in [Microsoft R Open] or not
  - `rstudio`     - (logical) whether running R in [RStudio] Console or not
  - `rstudioterm` - (logical) whether running R in [RStudio] Terminal or not
  - `wine`        - (logical) whether running R on Windows via [Linux Wine] or not

* Installed packages:

  - `package`     - (character) whether a package is installed or not.  In addition to checking the availability, having `package=<name>` in the filename makes it clear that the startup file concerns settings specific to that package.

* With a specific frequency:

  - `when` - (character) specify how often the file should be processed:
    - `when=once`        - the startup file is processed only once
    - `when=hourly`      - the startup file is processed at most once per hour
    - `when=daily`       - the startup file is processed at most once per day
    - `when=weekly`      - the startup file is processed at most once per week
    - `when=fortnightly` - the startup file is processed at most once every two weeks
    - `when=monthly`     - the startup file is processed at most once per month
    
  If such a file, or its timestamp, is updated, then it will be processed the next time R is started.

* Environment variables:

  - Any further `<key>=<value>` specifications with keys matching none of the above known keys are interpreted as system environment variables and **startup** will test such conditions against their values.  _Note, if `<key>` does not correspond to a known environment variable, then the file is skipped if `<key>=<value>` is used but included if `<key>!=<value>` is used._


To condition on more than one key, separate `<key>=<value>` pairs by commas, e.g. `~/.Rprofile.d/work,interactive=TRUE,os=windows.R`.  This also works for directory names.  For instance, `~/.Rprofile.d/os=windows/work,interactive=TRUE.R` will be processed if running on Windows and in interactive mode.  Multiple packages may be specified.  For instance, `~/.Rprofile.d/package=devtools,package=future.R` will be used only if both the devtools and the future packages are installed.

It is also possible to negate a conditional filename test by using the `<key>!=<value>` specification.  For instance, `~/.Rprofile.d/package=doMC,os!=windows.R` will be processed if package `doMC` is installed and the operating system is not Windows.


### Secrets (conditionally on an environment variable)

Renviron and Rprofile startup files with a non-declared `<key>` in their file names are skipped.  A non-declared key is any key that is neither one of the above predefined keys nor a declared environment variable.

This is useful, because it allows us to keep "secrets" in private files and only load them conditionally on the value of an environment variable.  For instance, we put all our secret Renviron files under `~/.Renviron.d/private/SECRET=banana/` such that they are only loaded if environment variable `SECRET` is set to exactly `banana`.  Moreover, by making sure the directory `~/.Renviron.d/private/` is only accessible by you, then it will be harder for someone else two know what your secret "password" (`banana`) is.  On a Unix system, you can set this up as:

```sh
$ mkdir -p ~/.Renviron.d/private/SECRET=banana/
$ chmod go-rw ~/.Renviron.d/private
```

Then go ahead an put your secret Renviron files in there, e.g.

```sh
.Renviron.d/private/SECRET=banana/
  +-- amazon  ## AWS_* credential & settings
  +-- github  ## GITHUB_TOKEN + GITHUB_PAT
```

Now, only if `SECRET` is set to exactly `banana` when you launch R, the above secret files will be processed and the corresponding environment variables will be available from withing R.  For instance, they will be loaded if you do:

```r
$ SECRET=banana R
```

but not if you do:

```r
$ SECRET=badguess R
```

or if `SECRET` is unset.

_Comment:_ You can use whichever variable name you like, it does not have to be `SECRET`.  And, the "password" `banana` is obviously just an example.


## Known limitations

### Setting environment variables during startup

Renviron startup files is a convenient and cross-platform way of setting environment variables during the R startup process.  However, for some of the environment variables that R consults must be set early on in the R startup process (immediately after Step 1), because R only consults them once.  Examples(*) of environment variables that need to be set _no later than_ `.Renviron` (Step 1) are:

* `TMPDIR`, `TMP`, `TEMP` - the parent of R's temporary directory, cf. `?tempdir`
* `LC_ALL` - locale settings used by R, e.g.  cf. `?locales`
* `R_DEFAULT_PACKAGES` - default set of packages loaded when R starts, cf. `?Rscript`
* `R_LIBS_USER` - user's library path, e.g. `R_LIBS_USER=~/R/%p-library/%v` is the folder specification used by default on all platforms and and R version.  The folder must exist, otherwise it is ignored by R.  The `%p` (platform) and `%v` (version) parts are R-specific conversion specifiers, cf. `?R_LIBS_USER`
* `R_MAX_NUM_DLLS`, cf. `?dyn.load`

Any changes to these done in an 'Renviron.d' file (Step 3a), or via `Sys.setenv()` in `.Rprofile` (Step 2) or 'Rprofile.d' files (Step 3c), _will be ignored by R itself_ - despite being reflected by `Sys.getenv()`.

Furthermore, some environment variables can not even be set in `.Renviron` (Step 1) but must be set _prior_ to launching R.  This is because those variables are consulted by R very early on (prior to Step 1).  Examples(*) of environment variables that need to be set _prior to_ `.Renviron` (Step 1) are:

* `HOME` - the user's home directory
* `MKL_NUM_THREADS` and `OMP_NUM_THREADS` - the default number of threads used by OpenMP etc, cf. _R Installation and Administration_
  
These variables have to be set using methods specific to the operating system or the calling shell, e.g. in a Unix shell

```sh
$ export OMP_NUM_THREADS=1
$ R
```

or per call as

```sh
OMP_NUM_THREADS=1 R
```

(*) For further details on which environment variables R consults and what they are used for by R, see the R documentation and help, e.g. `?"environment variables"` and `?Startup`.


## Examples
Below is a list of "real-world" example files:

```
.Renviron.d/
  +-- lang
  +-- r_cmd_check
  +-- secrets=banana
 
.Rprofile.d/
  +-- interactive=TRUE/
      +-- help.start.R
      +-- misc.R
	  +-- package=fortunes.R
  +-- os=windows.R
  +-- repos.R
```

They are available as part of this package under `system.file(package = "startup")`, e.g.

```r
> f <- system.file("Rprofile.d", "repos.R", package = "startup")
> file.show(f, type = "text")

local({
  repos <- c(CRAN = "https://cloud.r-project.org")
  if (.Platform$OS.type == "windows") {
     repos["CRANextra"] <- "https://www.stats.ox.ac.uk/pub/RWin"
  }
  options(repos = c(repos, getOption("repos")))
})
```

[startup]: https://cran.r-project.org/package=startup
[Emacs Speaks Statistics (ESS)]: https://ess.r-project.org/
[Microsoft R Open]: https://mran.microsoft.com/open
[pqR]: http://www.pqr-project.org/
[radian]: https://github.com/randy3k/radian
[RStudio]: https://www.rstudio.com/products/RStudio/
[Linux Wine]: https://www.winehq.org/

## Installation
R package startup is available on [CRAN](https://cran.r-project.org/package=startup) and can be installed in R as:
```r
install.packages("startup")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/startup", ref="develop")
```
This will install the package from source.  

<!-- pkgdown-drop-below -->



NEWS
Package: startup
================

Version: 0.17.0-9008 [2022-04-26]

NEW FEATURES:

 * Now startup::startup(debug = TRUE) reports on any environment
   variables, options, global variables, locale settings that were
   added, removed, and changed while processing an Rprofile file.

 * startup::startup(debug = TRUE) now reports on any changes to the
   current working directory and the R library path while processing
   an Rprofile file.

 * startup::startup(debug = TRUE) already reported on updates to the
   state of R's random number generator (RNG). Now it also reports on
   updates to the RNG kind, including what the change was.
   

Version: 0.17.0 [2022-02-23]

NEW FEATURES:

 * startup::startup() now searches for Renviron.d/ and Rprofile.d/ (note,
   without leading periods) in user's configuration folder in a way that is
   compatible with tools::R_user_dir() of R (>= 4.0.0) and that follows
   operating-system standards. For example, for Linux ~/.config/R/startup/,
   for macOS ~/Library/Preferences/org.R-project.R/R/startup/, and for
   MS Windows %LOCALAPPDATA%/R/cache/R/startup/. If environment variable
   'XDG_CONFIG_HOME' is set, then ${XDG_CONFIG_HOME}/R/startup/ is used.
   For now, startup::install() will continue to create ~/.Renviron.d/ and
   ~/.Rprofile.d/ by default.

 * startup::install() gained argument 'make_dirs' to control whether
   directories .Renviron.d/ and .Rprofile.d/ should be created, if missing.
   
 * Add on_session_enter() for registering R functions and expressions to be
   evaluated at the end of R's startup process via a custom .First() that is
   added to the global environment.  If there is a .First() function on the
   search path, including any preexisting one in the global environment, that
   is called at the end.

 * Add on_session_exit() for registering R functions and expressions to be
   called at the very end when the R session terminates.

 * Now startup() warns about non-standard, platform-specific capitalization of
   Renviron and Rprofile file names.  For example, although ~/.RProfile works
   on MS Windows and macOS with non-case sensitive file systems, the officially
   supported file name is ~/.Rprofile.

 * Now startup(debug = TRUE) reports on '.Last()' and '.Last.sys()'.

 * Now startup(debug = TRUE) gives a note if it detects that an Rprofile
   script updated the state of R's random number generator (RNG).

 * Harmonized cache folder options toward tools::R_user_dir() of R (>= 4.0.0).

BUG FIXES:

 * startup(unload = TRUE) failed when there was an .RData file and environment
   variable 'R_STARTUP_RDATA' or R option 'startup.rdata' specifies "prompt".



Version: 0.16.0 [2021-11-19]

NEW FEATURES:

 * Now startup(debug = TRUE) reports on 'R_SCRIPT_DEFAULT_PACKAGES'.

 * Now startup(debug = TRUE) reports on tempdir() and environment variables
   'TMPDIR', 'TMP', and 'TEMP' that controls the root of the temporary folder.

 * Now it is sufficient to call startup(dryrun = TRUE) to see what the R
   startup would have been.  Previously one would also have to specify
   debug = TRUE.

 * Now startup(debug = TRUE) reports on packages loaded, attached, unloaded,
   and detached per Rprofile file processed.  It also reports on other
   environments attached and detached.

 * Now startup(debug = TRUE) reports also on the Rprofile file in the 'base'
   package, i.e. system.file("R", "Rprofile", package = "base").

 * startup::startup(check = TRUE) no longer warns if environment variables
   'R_LIBS_SITE' or 'R_LIBS_USER' have value 'NULL', which in R (>= 4.2.0)
   corresponds to setting them to be empty.
   
BUG FIXES:

 * startup::restart(quiet = TRUE) would give an error in 'radian' saying
   '--quiet' is not supported, but it's supported in radian (>= 0.2.8)
   since 2018-08-27.
    

Version: 0.15.0 [2020-09-03]

NEW FEATURES:

 * startup(debug = TRUE) now reports on number of lines and file size also
   for the 'R_TESTS' file.

 * startup(debug = TRUE) [sic!] now checks that the environment variables that
   the different Renviron files set are actually set and produce an informative
   warning message if not. This confirms that, at least for R 4.0.2 for Windows,
   Rgui does *not* process ~/.Renviron unless Rgui is launched from that folder.

 * Using regular, non-fancy single quotes in any output produced.

BUG FIXES:

 * Package tests could attempt to backup and update ~/.Rprofile, e.g. if
   it did not have a newline on the last line.


Version: 0.14.1 [2020-03-31]

NEW FEATURES:

 * Add R options as an alternative to corresponding environment variables,
   e.g. added 'startup.disable' ('R_STARTUP_DISABLE'), 'startup.init'
   ('R_STARTUP_INIT'), and 'startup.rdata' ('R_STARTUP_RDATA').

DOCUMENTATION:

 * Add help("startup.options", package = "startup"), which lists environment
   variables and R options that the 'startup' package use.

BUG FIXES:
   
 * The new default of option 'stringsAsFactors' in R (>= 4.0.0) would trigger
   a startup warning.

 * It was not possible to disable the validation of R option 'error' and
   environment variables such as 'R_LIBS_USER', 'R_PROFILE_SITE', and
   'R_CHECK_ENVIRON'.  These validations can now be disabled by specifying
   argument 'check = FALSE', or the corresponding R option, or environment
   variable.  Validation of 'error' specifically can be disable by adding
   'error' to environment variable 'R_STARTUP_CHECK_OPTIONS_IGNORE' or to
   option 'startup.check.options.ignore'. The default is to ignore 'error'.


Version: 0.14.0 [2019-12-09]

SIGNIFICANT CHANGES:

 * startup(all = TRUE) now processes any startup folders in the home directory
   _before_ those in the current working directory.  Previously it was vice
   verse. With the new way, it is possible to override settings made in
   ~/.Renviron.d/ or ~/.Rprofile.d/ with custom ones in local ./.Renviron.d/
   and ./.Rprofile.d/ folders.

NEW FEATURES:


R/check.R
#' Check for and fix common mistakes in .Rprofile
#'
#' Check for and fix common mistakes in \file{.Rprofile} files.
#;
#' @param all Should all or only the first entry on
#' [the R startup search path][base::Startup] be checked?
#'
#' @param fix If `TRUE`, detected issues will be tried to be automatically
#' fixed, otherwise not.
#'
#' @param backup If `TRUE`, a timestamped backup copy of the original file is
#' created before modifying it, otherwise not.
#'
#' @param debug If `TRUE`, debug messages are outputted, otherwise not.
#'
#' @return Returns invisibly a character vector of files that were "fixed"
#' (modified), if any.  If no files needed to be fixed, or `fix = TRUE`,
#' then an empty vector is returned.
#' 
#' @references
#' 1. R-devel thread 'Last line in .Rprofile must have newline (PR#4056)',
#'    2003-09-03,
#'    \url{https://stat.ethz.ch/pipermail/r-devel/2003-September/027457.html}
#'
#' @export
check <- function(all = FALSE, fix = TRUE, backup = TRUE, debug = FALSE) {
  debug(debug)
  
  updated <- check_rprofile_eof(all = all, fix = fix, backup = backup,
                                debug = debug)
  
  check_rprofile_update_packages(all = all, debug = debug)
  
  if (!fix) {
    log("All startup files checked. If there were files with issues, they were not corrected because fix = FALSE.")
  } else {
    if (length(updated) == 0L) {
      log("All startup files checked. No files were fixed.")
    } else {
      logf("All startup files checked. The following files were fixed (modified): %s", paste(squote(updated), collapse = ", "))
    }
  }
  
  invisible(updated)
}


check_rprofile_eof <- function(files = NULL, all = FALSE, fix = TRUE,
                               backup = TRUE, debug = FALSE) {
  updated <- character(0L)

  debug(debug)
  if (is.null(files)) files <- find_rprofile(all = all)

  for (kk in seq_along(files)) {
    file <- files[kk]
    if (!eof_ok(file)) {
      if (fix) {
        if (backup) backup(file)

        ## Try to fix it by appending a newline
        try(cat(file = file, "\n", append = TRUE))

        ## Record that the file was updated
        updated <- c(updated, file)
        
        if (eof_ok(file)) {
          msg <- sprintf("SYNTAX ISSUE FIXED: Added missing newline to the end of file %s, which otherwise would cause R to silently ignore the file in the startup process.", file)  #nolint
          warning("startup::check(): ", msg)
        } else {
          msg <- sprintf("SYNTAX ERROR: Tried to add missing newline to the end of file %s, which otherwise would cause R to silently ignore the file in the startup process, but failed.", file)  #nolint
          stop("startup::check(): ", msg)
        }
      } else {
        msg <- sprintf("SYNTAX ERROR: File %s is missing a newline at the end of the file, which most likely will cause R to silently ignore the file in the startup process.", file)  #nolint
        stop("startup::check(): ", msg)
      }
    }
  }

  invisible(updated)
}


check_rprofile_update_packages <- function(files = NULL, all = FALSE,
                                           debug = FALSE) {
  if (is.null(files)) {
    files <- find_rprofile(all = all)
    paths <- find_rprofile_d(all = all)
    files <- c(files, list_d_files(paths))
  }
  if (length(files) == 0) return()

  patterns <- c(
    "utils::update.packages()" = "update[.]packages[(][^)]*[)]",
    "pacman::p_up()" = "p_up[(][^)]*[)]"
  )

  for (file in files) {
    bfr <- readLines(file, warn = FALSE)
    bfr <- gsub("#.*", "", bfr, fixed = FALSE)

    for (name in names(patterns)) {
      pattern <- patterns[name]
      if (any(grepl(pattern, bfr, fixed = FALSE))) {
        msg <- sprintf("UNSAFE STARTUP CALL DETECTED (%s): Updating or installing R packages during R startup will recursively spawn off an infinite number of R processes. Please remove offending call in order for .Rprofile scripts to be applied: %s", name, file)  #nolint
        stop("startup::check(): ", msg)
      }
    }
  }
}


check_options <- function(include = c("encoding", "error", "stringsAsFactors"), exclude = NA) {
  include <- match.arg(include, several.ok = TRUE,
                       choices = c("encoding", "error", "stringsAsFactors"))
  if (length(include) == 0L) return()
  if (length(exclude) > 0L) {
    if (is.na(exclude)) {
      ignore <- Sys.getenv("R_STARTUP_CHECK_OPTIONS_IGNORE", "error")
      if (is.na(ignore)) ignore <- NULL
      exclude <- getOption("startup.check.options.ignore", ignore)
    }
    keep <- (match(include, table = exclude, nomatch = 0L) == 0L)
    include <- include[keep]
  }
  if (length(include) == 0L) return()

  msg <- function(opt, default, value, body = NULL) {
    msg <- sprintf("R option '%s' was changed (to '%s') during startup, cf. Startup.  Values other than the default '%s' is known to cause problems.", opt, value, default)
    msg <- c(msg, body)
    msg <- c(msg, sprintf("To disable this check, add \"%s\" to option 'startup.check.options.ignore'.", opt))
    paste("startup::check():", paste(msg, collapse = " "))
  }

  for (opt in include) {
    if (opt == "encoding") {
      value <- getOption(opt, default)
      default <- "native.enc"
      if (!interactive() && value != default) {
        warning(msg(opt, default, value, body = "For example, in non-interactive sessions installation of packages with non-ASCII characters (also in source code comments) fails. To set the encoding only in interactive mode, e.g. if (base::interactive()) options(encoding = \"UTF-8\")."), call. = FALSE)
      }
    } else if (opt == "error") {
      check_rstudio_option_error_conflict()
    } else if (opt == "stringsAsFactors") {
      value <- getOption(opt, default)
      default <- if (getRversion() >= "4.0.0") FALSE else TRUE
      if (value != default) {
        warning(msg(opt, default, value), call. = FALSE)
      }
    }
  }
}


check_envs <- function() {
  check_r_libs_env_vars()
}


check_r_libs_env_vars <- function() {
  vars <- c("R_LIBS", "R_LIBS_SITE", "R_LIBS_USER")
  for (var in vars) {
    path <- Sys.getenv(var)
    if (!nzchar(path)) next

    ## Ignore "NULL" as is the case in R 4.2.0?
    if (var != "R_LIBS") {
      if (path == "NULL") next
    }

    ## Don't check intential "dummy" specification, e.g.
    ## non-existing-dummy-folder
    is_dummy <- grepl("^[.]", path) && !grepl("[/\\]", path)
    if (is_dummy) next
    
    paths <- unlist(strsplit(path, split = .Platform$path.sep, fixed = TRUE))
    paths <- unique(paths)
    paths <- paths[!vapply(paths, FUN = is_dir, FUN.VALUE = FALSE)]
    npaths <- length(paths)
    if (npaths > 0) {
      pathsx <- normalizePath(paths, mustWork = FALSE)
      pathsq <- paste(squote(paths), collapse = ", ")
      pathsQ <- paste(sprintf("\"%s\"", paths), collapse = ", ")
      pathsxq <- paste(squote(pathsx), collapse = ", ")
      if (npaths == 1L) {
        msg <- sprintf("Environment variable %s specifies a non-existing folder %s (expands to %s) which R ignores and therefore are not used in .libPaths(). To create this folder, call dir.create(%s, recursive = TRUE)", squote(var), pathsq, pathsxq, pathsQ)
      } else {
        msg <- sprintf("Environment variable %s specifies %d non-existing folders %s (expands to %s) which R ignores and therefore are not used in .libPaths(). To create these folders, call sapply(c(%s), dir.create, recursive = TRUE)", squote(var), npaths, pathsq, pathsxq, pathsQ)
      }
      warning("startup::check(): ", msg, call. = FALSE)
    }
  }

  vars <- c("R_PROFILE", "R_PROFILE_USER", "R_ENVIRON", "R_ENVIRON_USER")
  for (var in vars) {
    pathname <- Sys.getenv(var)
    if (!nzchar(pathname)) next
    
    if (!is_file(pathname)) {
      pathnamex <- normalizePath(pathname, mustWork = FALSE)
      msg <- sprintf("Environment variable %s specifies a non-existing startup file %s (expands to %s) which R will silently ignore",
                     squote(var), squote(pathname), squote(pathnamex))
      warning("startup::check(): ", msg, call. = FALSE)
    }
  }

  vars <- c(build = "R_BUILD_ENVIRON", check = "R_CHECK_ENVIRON")
  for (key in names(vars)) {
    var <- vars[key]
    pathname <- Sys.getenv(var)
    if (!nzchar(pathname)) next
    
    if (!is_file(pathname)) {
      pathnamex <- normalizePath(pathname, mustWork = FALSE)
      msg <- sprintf("Environment variable %s specifies a non-existing startup file %s (expands to %s) which 'R CMD %s' will silently ignore",
                     squote(var), squote(pathname), squote(pathnamex), key)
      warning("startup::check(): ", msg, call. = FALSE)
    }
  }
}


check_rstudio_option_error_conflict <- function() {
  ## Nothing to do?
  if (is.null(getOption("error")) || !is_rstudio_console()) return()

  ## Skip check if renv is active (because we cannot reliably test)
  ## https://github.com/HenrikBengtsson/startup/issues/76
  is_renv <- isTRUE(as.logical(Sys.getenv("RENV_R_INITIALIZING")))
  if (is_renv) return()

  ## If possible, detect when 'Debug -> On Error' is _not_ set in RStudio.
  ## If so, then skip the warning, because that is a case when RStudio Console
  ## does not override 'error'.
  config_root <- "~/.rstudio-desktop"
  if (!is_dir(config_root) && sysinfo()$os == "windows") {
    ## Officially documented root folder for RStudio configuration files
    ## Source: https://support.rstudio.com/hc/en-us/articles/200534577-Resetting-RStudio-Desktop-s-State
    
    ## Alternatives on (a) Windows Vista 7, 8, ... and (b) Windows XP
    config_root <- file.path(Sys.getenv("localappdata"), "RStudio-Desktop")
    if (!is_dir(config_root)) {
      config_root <- file.path(Sys.getenv("USERPROFILE"), "Local Settings",
                               "Application Data", "RStudio-Desktop")
    }
  }
  if (is_dir(config_root)) {
    ## Non-official configuration file found by reverse engineering only,
    ## cf. https://github.com/HenrikBengtsson/startup/issues/59
    config_file <- file.path(config_root, "monitored", "user-settings",
                             "user-settings")
    if (is_file(config_file)) {
      config <- readLines(config_file, warn = FALSE)
      ## 'Debug -> On Error' is _not_ set.  Nothing to warn about
      config <- grep("errorHandlerType=", config, fixed = TRUE, value = TRUE)
      if (length(config) == 0L) return()
      if (any(grepl("errorHandlerType=\"3\"", config, fixed = TRUE))) return()
    }
  }

  ## Record intended value of option 'error'
  options(startup.error.lost = getOption("error"))

  warning("startup::check(): ", "CONFLICT: Option ", squote("error"), " was set during the R startup, but this will be overridden due to the RStudio settings (menu ", squote("Debug -> On Error"), ") when using the RStudio Console. To silence this warning, do not set option 'error' when running RStudio Console, e.g. ", squote("if (!startup::sysinfo()$rstudio) options(error = ...)"), ". The 'error' option that was set during the startup process but lost is recorded in option ", squote("startup.error.lost"), ". For further details on this issue, see https://github.com/rstudio/rstudio/issues/3007")
}


## Check that Renviron and Rprofile files are properly capitalized. The proper
## way is .Renviron and .Rprofile, whereas, for instance, .REnviron is not.
warn_file_capitalization <- function(pathname, what) {
  ## Get the actual name on file
  path <- dirname(pathname)
  file <- basename(pathname)
  pattern <- sprintf("^%s$", file)
  actual <- dir(path = path, pattern = pattern, ignore.case = TRUE, all.files = TRUE)
  if (length(actual) == 0) return(invisible(TRUE))
  if (length(actual) > 1) {
    warning(sprintf("startup::startup(): Unexpected, internal result. Please report to the %s maintainer. Found more than one %s file: %s", squote(.packageName), squote(what), paste(squote(actual), collapse = ", ")))
    actual <- actual[1]
  }
  
  ## Is it a non-standard file name?
  pattern <- sprintf("^[.]?%s", what)
  if (grepl(pattern, actual)) return(invisible(TRUE))

  ## Produce informative warning
  pathname_actual <- file.path(path, actual)
  correct <- gsub(pattern, sprintf(".%s", what), actual, ignore.case = TRUE)
  pathname_correct <- file.path(path, correct)
  msg <- sprintf("Detected non-standard, platform-dependent letter casing of an %s file. Please rename file to use the officially supported casing: %s -> %s", squote(what), squote(pathname_actual), squote(pathname_correct))
  warning("startup::startup(): ", msg, call. = FALSE)
  invisible(FALSE)
}
